Update (#12)
vacuumlabs/coding-tips: src/index.adoc

- link:https://www.infoq.com/presentations/Are-We-There-Yet-Rich-Hickey[Are we
  there yet^]

This is Rich (again!) promoting immutable values and immutable data structures
over references that change over time in a mutable manner. Also, Rich talks
about the Epochal time model, where the whole (!) state of the program is
represented by a single mutable reference that takes a sequence of immutable
values, modeling a time-evolution of a state. This model is used with Redux
framework.

=== link:https://www.youtube.com/watch?v=FihU5JxmnBg[Stuart Halloway: Debugging
with the Scientific Method^]

Be less chaotic and more organized when debugging. State explicitly what's your
hypothesis, test it, think. Write things down. Test one thing at a time. A good
process is more important than tools; scientific approach with console.log is
more powerful than ad-hoc trying random stuff with a strong debugger.

== Tools

=== Operating System
Use either Mac or Linux for development. Don't use Windows, don't even think
about it! There is always something working differently on Windows than on the
other platforms. This costs us huge amounts of time debugging problems instead

> of developing valuable features.

Reviewer: Well. I'm not sure it counts, quite a few people I know work from WSL = windows subsystem for linux. You basically get a virtualized linux inside of Windows, but the layer is very lightweight and you don't loose too much performance. Big bonus is that you avoid most of the driver issues.

Maybe would change it to "native" Windows, or mention that WSL2 is actually fine. I'm not saying I would recommend it over mac or linux, but it's much better than wasting days on broken laptop drivers and bad unstable UX that's often the case on modern linux desktop environments.

Reviewer 2: > ... broken laptop drivers ...

No issues with Intel and AMD gpus. I'm not gaming anymore, though. For office usage it works smoothly.

> ... bad unstable UX that's often the case on modern linux desktop environments. ...

The "bad UX" is quite subjective, it more depends on habits. I'm fluently working with GNOME 3 with few extensions, and keybindings to switch to other workspaces.

As of the "unstable" part, well I haven't seen this since I moved from ArchLinux to Ubuntu and I'm using only LTS releases.

Reviewer: It's not the gpus I worry about too much, that gets enough visibility. I haven't had a notebook that didn't have issues with at least one of these things: wifi, bluetooth, soundcard, sleep, touchpad, thunderbolt, webcam, keyboard shortcuts, touchpad gestures (multitouch), touchscreen or webcam (although lately the list is usually shrinking). Our lenovos I think are doing quite OK though apart from touch :).

If you are new to Linux, install link:http://www.ubuntu.com[Ubuntu^]. Make sure to read
link:https://help.ubuntu.com/stable/ubuntu-help[the introduction to Ubuntu Desktop^], especially

> link:https://help.ubuntu.com/stable/ubuntu-help/addremove.html[how the software installation work in Linux^].

Reviewer: I would even go as far as recommend manjaro over Ubuntu for newcomers. Better installation, better package manager with more packages out of the box. The default DEs KDE/XFCE >>> gnome for anyone new to linux. (gnome a very unintuitive UX for anyone coming from windows)

Reviewer 3: Well, I would recommend against Manjaro. It's a rolling distro and don't forget that can be often tricky even for power users. I am, well, not a Linux guru but a long time user that tried many distros.

Manjaro is maybe for example a good gaming distro but in my year-long experience it got broken a few times and I had to fix it manually.
There is AUR which is good for users like me but stuff there is often broken too and one has to edit install scripts. Moreover the packages there work against Arch packages and Manjaro is 1-2 weeks late compared to Arch - and that breaks stuff too.

Yes, many of us are against Ubuntu because of various reasons but hey, it's still a good, "simple" distro for beginners together with Mint.
They can easily switch to their own choice when they would be capable of doing that.

Reviewer 2: I had been using ArchLinux in past, and there was some amount of time spent due to things being broken by updates. Since I moved to Ubuntu, maintenance costs / time spent was heavily reduced. I am not heavily chasing performance anymore, though.

If I was to switch to something else, which would be more complex than Ubuntu, then I would consider Qubes OS. It would give me more value (increased security) for the extra maintenance cost. It might even reduce costs in long-term, as it allows to separate project specific tools, and easily drop them completely (delete qube / virtual machine) without a need to reinstall OS to be sure there aren't any leftovers.

// ==== Linux

==== macOS

Get a _i_**Mac**_Book_ _Pro_. macOS can run on non-Apple computer, it's called
Hackintosh, but things can work differently there and it's against EULA. 😞

macOS has integrated App Store from which you can download some useful
applications like XCode, Slack, and iWork (Pages, Numbers and Keynote) apps that
came with your new Mac. You will use App Store to update the system, too.

macOS doesn't come with package manager as is `apt` in the Ubuntu. There are two
options link:https://brew.sh[Homebrew^] and
link:https://www.macports.org[MacPorts^]. The most used in our company is
Homebrew.

=== Terminal

Learn link:http://linuxcommand.org/learning_the_shell.php[the basics of UNIX shell^].

Replace your default shell with much better one called _zsh_. Install

> link:https://ohmyz.sh/[oh-my-zsh^], where you can select some nice visual theme for your shell.

Reviewer: Imho for beginners, bash should be the standard. Most guides and tools work really well with that. With zsh even on mac, a lot of tools have hiccups, bad installations, bad env setup. On linux it made me switch away from it. Last time we ran into that was with `volta.sh` . You can install powerline for bash as well.
https://github.com/powerline/powerline
https://medium.com/earlybyte/powerline-for-bash-6d3dd004f6fc

I shared some stuff I'm (Tomas Kulich) using. A plugin called
link:https://gist.github.com/tomaskulich/2957cba62b241d362a0ea3324ea2f2f4[swissknife^] to make shell
more user friendly, my
link:https://gist.github.com/tomaskulich/1d212e6c9a9381f7f25368e519dc87ab[zsh theme]. I want to see more
than just a current folder's name on my prompt, but not the whole path either. Highly recommended is
also `git-prompt` plugin to give you better git status to your prompt (such as: how many
commits before and ahead of the remote my branch is.)

If you know any other cool tricks, please contribute!

=== Text Editor

==== Visual Studio Code

image::editors/screenshot-vscode.png[]

link:https://code.visualstudio.com[Visual Studio Code site]

Open source editor by Microsoft. It comes prepared for web-development with

> integrated support for JavaScript intellisense, git integration and with many

Reviewer: +Typescript

global config `~/.gitconfig`
```
[user]
     name = Joe Srna
     email = joe.srna@gmail.com
[includeIf "gitdir:~/Projects/vacuumlabs/"]
    path = ~/Projects/vacuumlabs/.gitconfig
```

work related config `~/Projects/vacuumlabs/.gitconfig`
```
[user]
     email = joe.srna@vacuumlabs.com
```

You can verify the change by running `git config --list` within different directories.
===============================

== Tutorials

Learning HTML/CSS::
link:http://www.w3schools.com[W3 Schools] is a great starter point for learning
basics of HTML and CSS.

Learning JavaScript::
Definitely, have a look at an introduction tutorial into JavaScript on
link:https://www.codecademy.com/learn/javascript[Codecademy]. Use
link:https://developer.mozilla.org/en-US/docs/Web/JavaScript[Mozilla Developer
Network] as your online JavaScript reference.

Learning JS React & Redux::
We use cool modern JavaScript technology like
link:http://facebook.github.io/react/[React] and link:http://redux.js.org[Redux].

Learning Python::
We use link:http://www.diveintopython3.net[Python3] with Django.
Have a look at this great link:http://gettingstartedwithdjango.com[Django tutorial].

Clojure intro::

> http://www.braveclojure.com/foreword/

Reviewer: We should replace this with Kotlin probably.

outside world. Furthermore, pure function shouldn't mutate its arguments and should use solely the
return statement to communicate the result to the caller.

Pure function can be tested and reasoned about in isolation. Therefore the second code snippet is a
way more preferred to the first:

```
// don't do this
function main() {
  function this_is_hard_to_test(filename) {
    const data = read_data_from_file(filename)
    do_something_with_data
    return something
  }
  this_is_hard_to_test('data.txt')
}

// do this
function main() {
  const data = read_data_from_file('data.txt')
  function this_is_better(data) {
    do_something_with_data
    return something
  }
  this_is_better(data)
}
```

Don't get trapped by dogma here: many tasks can not be easily addressed using only 100% pure functions.
Be ready to conduct some side-effects, read (and sometimes even modify) variables from the outer scopes,
etc. The learning here is, that you should pursue the purity to the reassonable level.

As an example, Haskell takes purity really seriously and perform side-effects by monads. Monads
seemingly make all functions pure, but in reality, it's the kind of purity which costs you a lot,

> while not buying you either better testability, nor it ease your resoning about the code.

Reviewer 4: ```suggestion
seemingly make all functions pure, but in reality, it's the kind of purity that costs you a lot -
while not buying you better testability, it doesn't make reasoning about the code easier either.
```

=== Prefer simple data structures

> In Java-ish languages one does often define structures like this:

Reviewer 4: ```suggestion
In Java-ish languages one often defines structures like this:
```

```
class Person {
  first_name: String
  last_name: String
  function full_name() : String {
    return this.first_name + ' ' + this.last_name
  }
}

const me = new Person(...)
```

In JS, TS, Clojure the second approach is very common:

```

> const me:User = {first_name: "Tomas", last_name: "Kulich"}

Reviewer 4: ```suggestion
const me: User = {first_name: "Tomas", last_name: "Kulich"}
```

implement something that feels like a general object-manipulation stuff", pause and reflect on your design.

=== Immutable data structures

Think of a typical String or Date implementations. Usually (Java, Python, JS), these are immutable
objects; one can construct new string such as `my_string + "!"` but there is no way mutating the
string, e.g. no `my_string.append("!")` exist. This brings a great relief when reasoning about Strings:
if I see `my_string` variable somewhere, I can trust that no matter who else keeps the reference to
the same string, the stuff is not going to change.

It's natural to strive for the immutability benefits also with Maps, Lists (and Objects). In JS
there are several easy ways how to update the object in an immutable manner:

```
const updated_user = {...user, name: new_name}
// or this one, ideal for nested objects
const updated_user = immutability_helper.update(user, {name: {$set: new_name}})
```
There exist immutable data structures implementation (Map, List) that do this in a very performant
manner (time complexity `O(lg(structure_size))`), however, unless the structures are huge, you can

> just stick to good-old mutable objects and just not mutating them by convention.

Reviewer 4: ```suggestion
just stick to the good-old mutable objects and just avoid mutating them by convention.
```

=== Stay pragmatic

There are many wrong things to optimize for:
- extreme purity,
- coding the idiomatic [your favorite framework] way,
- 100% test coverage,
- your code behaving 100% correctly even in unlike and unimportant situations, etc.

We're solving problems. Focus on "what's the problem we're trying to solve here" rather than "what's

> the most idiomatic way to achive this in framework X".

Reviewer 4: ```suggestion
the most idiomatic way to achieve this in framework X".
```

This doesn't advocate writing shitty code. Saving 30 minutes of your time just to create a day long
debug session for yourself is certainly not pragmatic.

=== Avoid premature optimization and premature generalizations

> Always be aware to how much resources (time, memory) you can waste, so you can waste them visely.

Reviewer 4: ```suggestion
Always be aware of how much resources (time, memory) you can waste, so you can waste them wisely.
```

With time, be aware of what's the most time-consuming thing that's happening and ignore everything

> else, prefering readability to unimportant time-saving. If a reviewer asks you to optimize O(n) work

Reviewer 4: ```suggestion
else, preferring readability to unimportant time-saving. If a reviewer asks you to optimize O(n) work
```

will be needed in 2 months, the prefered way  to achieve X is likely to change by that time. And
even if you're completely right about all your predictions, it'll still may be easier postpone X for
later, since you won't need to maintain X for 2 extra months.

=== Produce bug-free code by avoiding bugs

Imagine you write 100 lines. You hit run and--it's working! What a feeling! Some people do it
differently: they quickly produce 100 lines of bad code, then using print and debugger to get the
code in shape. Not only this feels worse, in the end, it's also much more time consuming.

If you fall into the second case, consider changing your workflow. Slow down and focus on writing
high-quality code. Reason why the code should work ("On this place, the connection must be already
opened, because I jump here only from here and here... wait, also from here! Oh my, that'd be a
nasty bug to have."). Play a "coding game": If the code runs correctly on the first time (possibly
after fixing small and easy bugs such as typos), you've won. Soon you develop a healthy ammount of
paranoia, which'll make you question your code frequently. You'll produce a code 3x slower but also

> 10x faster.

Reviewer 4: ```suggestion
Be very humble when predicting what new features the business will need. Even if you're right that X
will be needed in 2 months, the preferred way to achieve X is likely to change by that time. And
even if you're completely right about all your predictions, it still may be easier to postpone X for
later, since you won't need to maintain X for 2 extra months.

=== Produce bug-free code by avoiding bugs

Imagine you write 100 lines. You hit run and--it's working! What a feeling! Some people do it
differently: they quickly produce 100 lines of bad code, then using print and debugger to get the
code in shape. Not only this feels worse, in the end, it's also much more time-consuming.

If you fall into the second case, consider changing your workflow. Slow down and focus on writing
high-quality code. The reason why the code should work ("On this place, the connection must be already
opened, because I jump here only from here and here... wait, also from here! Oh my, that'd be a
nasty bug to have."). Play a "coding game": If the code runs correctly on the first try (possibly
after fixing small and easy bugs such as typos), you've won. Soon you develop a healthy amount of
paranoia, which'll make you question your code frequently. You'll produce a code 3x slower but also
10x faster.
```

=== Test are good, if they add value

How to write tests is almost a religious thingy, with several though schools. Stay away from the
dogmatic mess they've created; simple pragmatism is the best guidance. If the test adds value that
outweighs the cost (writing, maintaining and running the test), then it's a good test to have.

Imagine such a test:
- you populate a User object with data,
- save the obejct to the database,
- load an object from DB,
- assert that all attributes are correctly loaded.

The test has a low value and probably doesn't pull its weight:
- it's testing a relatively simple code; it'd be easier to review it manually,
- the code is not likely to break,
- if test passes, it doesn't mean much; there are many edge cases (e.g. very long user name), that are
  not covered,
- if test fails, you'd probably find out anyway really,

> - it doesn't serve as a documentation to something worth documenting.

Reviewer 4: ```suggestion
How to write tests is almost a religious thingy, with several thought schools. Stay away from the
dogmatic mess they've created; simple pragmatism is the best guidance. If the test adds value that
outweighs the cost (writing, maintaining, and running the test), then it's a good test to have.

Imagine such a test:
- you populate a User object with data,
- save the object to the database,
- load an object from DB,
- assert that all attributes are correctly loaded.

The test has a low value and probably doesn't pull its weight:
- it's testing a relatively simple code; it'd be easier to review it manually,
- the code is not likely to break,
- if the test passes, it doesn't mean much; there are many edge cases (e.g. very long user name), that are
  not covered,
- if the test fails, you'd probably find out anyway really,
- it doesn't serve as documentation to something worth documenting.
```

=== Code reviews are awesome

They are the best tool we have to produce a bug-free, readable and maintainable code. Often, more
than one review is necessary for a code to be merged.

Few tips:

- Don't enforce your taste to others too much. If you know shorter form of how to write something

>   but you're not sure, just give advice: 'consider rewriting to ...`.

Reviewer 4: `'` vs `` ` ``

- This sucks: correcting 10 OKish namings just because it's easy to do and missing a huge design
  flaw at the same time.
- If you've had an objective problem understanding something and author had to explain it to you, it

>   probablyu deserves a comment.

Reviewer 4: ```suggestion
  probably deserves a comment.
```

- If it's for a longer discussion, schedule a call or just pick up the phone and call the other
  person.

Finally, if it's not clear why the code works correctly, you shouldn't merge it. Author can use

> different ways to persuade you, she did a good job:

Reviewer 4: ```suggestion
different ways to persuade you she did a good job:
```

- adding explanatory comments,
- writing tests, demonstrating the desired functionality,
- structuring a code in a way, that certain behavior is enforced by design,
- Q: "hey, I'm not sure about this", A: "Me neither, but I've tested it manually and it seems to
  work. It's poorly documented, so it'll do for now I guess."

The code review process is not so easy to get right.

- Reviewer comments to the PR.
- Author reacts :thumbsup:, if he accepts the proposed change. If it's something more complex, the

>   discussion can take place. Avoid writing a lot it's inefficient and frustrating. Have a quick call

Reviewer 4: ```suggestion
  discussion can take place. Avoid writing a lot, it's inefficient and frustrating. Have a quick call
```

  instead.
- Reviewer checks that she likes the outcome by clicking "mark as resolved".

> - If reviewer has only cosmettic comments, he can approve the PR indicating he doesn't care that

Reviewer 4: ```suggestion
- If reviewer has only cosmetic comments, she can approve the PR indicating he doesn't care that
```

  much which suggestions author implements.

Bad practices:

- Leave the comment without any reaction.
- Resolving the reviewer's comment by the author in a case where author is not 100% sure he fixed it
  in a way reviewer intended.

The changes are pushed either as a new "fixing comments" commit, or the history is editted.
Adapt the customs of your project.

=== Microservices are often overused

There's a too big hype around microservices. Obviously, the world doesn't run as one monolithic

> service (and that's a good think). Since "big project" ==  "small world", it's not a good idea for a

Reviewer 4: ```suggestion
service (and that's a good thinking). Since "big project" == "small world", it's not a good idea for a
```

big project to run as a monolithic service either. Structuring the project to smaller microservices
helps--as long as the microservices have well-defined responsibilities and the cross-cutting
concerns are limitted.

> Unhealthy hype about microservices gave birth to systems with too much microservices and a lot of

Reviewer 4: ```suggestion
Unhealthy hype about microservices gave birth to systems with too many microservices and a lot of
```

incidental complexity. Keep the number of microservices small and create them only after you're
pretty certain that such a service makes sense; don't be affraid to start with monolithic design and
then carve-out a few services later.

=== Monorepo is a better idea than newcomers usually think

This is somewhat similar to microservices. The world isn't in one giant GIT repo, but there's a lot
of cost if your production requires running a code from 20 different repos. If you think "we're too
big for a simple repo", then consider the whole Google codebase lives in a single repository; they
have a lot of tooling built around it, and as your system evolves, you'll probably need to build
some custom tooling as well.

=== Fat code is good
When you start coding, you name your variables `i`, `j`, `k`. Then you code a project with 10k lines
and you discover using better names such as: `user`, `account`, `ammount` or `transfer()` is good
for your sanity.
Then you code a project with 100k lines, discovering you need even better names:
`account_sender_or_system_account`,

> `ammount_USD_before_VAT`, `try_transfer_with_SEPA()`. Often people use

Reviewer 4: ```suggestion
`amount_USD_before_VAT`, `try_transfer_with_SEPA()`. Often people use
```

short and imprecise names that kick ass badly, once the project grows.

Good type system can help you shorten the names of your variables and parameters, since a lot of
info is just to describe the type. This is not always possible, especially for functions you should
be quite verbose.

Using long names results in writing a fat code: simple expression such as `a = b + c + d` can overflow the
100 characters limit, which looks ugly and not that readable. This is usually less severe

> problem as one may think.

Reviewer 4: ```suggestion
100 characters limit, which looks ugly and not that readable. However, this is usually a less severe
problem than one may think.
```

=== Comments

John produces a code. During code review, Jane doesn't understand a certain line, which creates the
communication such as:

Jane: What if `user` parameter is not provided?
John: Actually, it's always provided.
Jane: Why is the parameter declared as optional then?
John: Ah, I see your point. It's optional just for the tests. Real code always puts the value there.
I've just checked and it's really like that.
Jane: OK then. Ship it.

Then the code gets merged. But can you spot the process flaw? Jane had a valid question and John
answered it. However, no one thought about 20 other team members that'll encounter that line of
code. They'll be as lost as Jane, without any clue, why the code is like that.

As a rule of thumb, when explaining more complected or suspicious code to the curious reviewer,
consider adding a comment such as:

```
// Parameter `user` is actually required. It's marked optional just because of tests.
```

Another case, that always deserves a comment: copy-pasting code from somewhere. Always add a link to

> where the code came from.

Reviewer 4: I love these two tips. I can't stress them enough myself.

=== Idempotency

> Imagine a service that allows you to create new tansactions as such:

Reviewer 4: ```suggestion
Imagine a service that allows you to create new transactions as such:
```

```

> external_service.add_new_transaction(from: "john", to: "jane", usd_ammount: 10)

Reviewer 4: ```suggestion
external_service.add_new_transaction({from: "john", to: "jane", usd_amount: 10})
```

```

To handle possible errors, you need to add some try-catch logic:

```
try {

>   external_service.add_new_transaction({from_account: "john", to_account: "jane", usd_ammount: 10})

Reviewer 4: ```suggestion
  external_service.add_new_transaction({from_account: "john", to_account: "jane", usd_amount: 10})
```

} catch (err) {
  // what now?
}
```

How would you handle the error? You may want to repeat the transaction:

```

> function try_send_transaction_with_repeat({from_account, to_account, usd_ammount, try_this_many_times}) {

Reviewer 4: ```suggestion
function try_send_transaction_with_repeat({from_account, to_account, usd_amount, try_this_many_times}) {
```

  // default value
  try_this_many_times = try_this_many_times || 10
  try {
    external_service_do_your_job
  } catch (err) {
    if (try_this_many_times > 0) {
      wait(10000)

>       try_send_transaction({from_account, to_account, usd_ammount, try_this_many_times - 1})

Reviewer 4: ```suggestion
      try_send_transaction({from_account, to_account, usd_amount, try_this_many_times - 1})
```

    } else {
      throw Error('sa mi neda poslat')
      // pun intended
    }
  }
}
```

If you call this with `try_this_many_times === 10`, you'll get 10 tries in 10s intervals and then
error is thrown. But what if try_send_transaction actually _created_ the transaction on the external
service and just then something failed, when sending the response back (e.g. parsing the response on
your server is faulty)? You'd end up with 10 transaction being sent: uh-oh.

To overcome this, decent external services let you specify so-called _idempotency key_ which serves
as unique identifiers to your request. If the external service is decent, you can use it as such:

```

> function try_send_transaction_with_repeat({from_account, to_account, usd_ammount, try_this_many_times, idempotency_key}) {

Reviewer 4: ```suggestion
function try_send_transaction_with_repeat({from_account, to_account, usd_amount, try_this_many_times, idempotency_key}) {
```

The changes are usually pushed as a new "fixing PR comments" commit to the feature branch. This
allows the reviewer to see the edits in the fixing commit, or see the whole proposed change in the
feature branch. For this to work, the feature branch shouldn't contain more than one logical change.
Also, the "fixing PR comments" commits are not useful after the branch is merged--therefore the
feature branch should be squashed to a small number of commits (often just one) before merging.

Another approach is: you can put more than one atomic change into a branch. In such a case, one
atomic change gets represented by one commit. The reviewer reviews the branch commit-by-commit and
when she requires change, author edits the history. Reviewer doesn't see the correcting diff itself,
but can then check out her comments, the current state of the code and see, if the code looks good
now. Pros: it works better for doing several tasks simultaneously. For example, you start coding a
feature, refactor something on-the-fly and fix a bug you discover. It's much easier to have just one
branch for that. Cons: reviewer doesn't see an exact diff, that author produced to implement their

> change.

Reviewer 5: > author edits the history. Reviewer doesn't see the correcting diff

A solution actually exists, `git range-diff`, though not many people seem to use it. I often find it useful in exactly this kind of situation: verifying some small change to a long commit series.
